packages/@react-spectrum/utils/src/styleProps.ts

```diff
@@ -249,3 +259,36 @@ export function useStyleProps<T extends StyleProps>(props: T, handlers: StyleHan
 export function passthroughStyle(value) {
   return value;
 }
+
+export function getResponsiveProp<T>(prop: Responsive<T>, breakpoint: Breakpoint): T {
+  if (prop && typeof prop === 'object') {
+    if (prop[breakpoint]) {
+      return prop[breakpoint];
+    } else {
+      return (prop as ResponsiveProp<T>).base;
+    }
+  }
+  return prop as T;
+}
+
+export function useBreakpoint(): Breakpoint {
```

Comment:
Perhaps this hook should go in `@react-spectrum/provider` rather than utils? That would avoid a circular dependency and I think it makes sense since the breakpoints are defined in provider as well.

An even better option might be to make this part of `useProvider` rather than a separate hook. Right now, each component that calls `useBreakpoint` creates and watches three media queries, so we could end up with a lot of duplicate media queries across a whole application. If we watched these media queries in the top-level provider instead, and just passed down the matched breakpoint as part of the provider context this would be reduced to just three media queries for the whole app.

Reply:
> An even better option might be to make this part of useProvider rather than a separate hook

I liked the idea and tried this.
Since `Provider` itself also consumes `useStyleProps`, 
the value of `breakpoint` from `useProvider` is undefined on top-level provider.
So I changed the type of `useStyleProps` to pass `breakpoint` directly to `useStyleProps`.

```tsx
// before
function useStyleProps<T extends StyleProps>(props: T, handlers: StyleHandlers) {}
```

```tsx
// after
function useStyleProps<T extends StyleProps>(props: T, options: {
  handlers?: StyleHandlers,
  breakpoint?: Breakpoint
}) {}
```

Reply:
and instead of using several `useMediaQuery`, 
I re-created a custom hook `useBreakpoint`, since we need to support custom breakpoints.
 
```ts
export function useBreakpoint(breakpoints: Breakpoints) {
  // sort breakpoints in ascending order.
  let entries = Object.entries(breakpoints).sort(([, valueA], [, valueB]) => valueA - valueB);
  let breakpointQueries = entries.map(([, value], index) => {
    if (index === entries.length - 1) {
      return `(min-width: ${value}px)`;
    } else {
      return `(min-width: ${value}px) and (max-width: ${entries[index + 1][1]}px)`;
    }
  });

  let supportsMatchMedia = typeof window !== 'undefined' && typeof window.matchMedia === 'function';
  let getBreakpointHandler = () => {
    let point = 'base';
    for (let i in breakpointQueries) {
      let query = breakpointQueries[i];
      if (window.matchMedia(query).matches) {
        point = entries[i][0];
        break;
      }
    }
    return point;
  };

  let [breakpoint, setBreakpoint] = useState(() =>
    supportsMatchMedia
      ? getBreakpointHandler()
      : 'base'
  );

  useEffect(() => {
    if (!supportsMatchMedia) {
      return;
    }

    let onResize = () => {
      setBreakpoint(getBreakpointHandler());
    };

    window.addEventListener('resize', onResize);
    return () => {
      window.removeEventListener('resize', onResize);
    };
  }, [supportsMatchMedia]);

  // If in SSR, the media query should never match. Once the page hydrates,
  // this will update and the real value will be returned.
  let isSSR = useIsSSR();
  return isSSR ? 'base' : breakpoint;
}
```

Reply:
Since `useStyleProps` and `useBreakpoint` changed, 
we need to change every component that consumes `useStyleProps`to avoid circular dependencies.
(by now, `useStyleProps` is consuming `useProvider` to use breakpoint from provider context.)

@devongovett 
Would you mind if I change a bunch of files that consume `useStyleProps`? (66 files)

Reply:
Ah yeah... hmm kinda annoying that we'd need to pass in the breakpoint everywhere we use style props. I think we need to somehow keep the breakpoint code in utils. I see two possibilities:

1. We could create a `BreakpointProvider` inside utils that is a context provider of the current breakpoint. Then `useStyleProps` could consume this. We'd need to then render a `BreakpointProvider` from inside the `Provider` component.
2. Alternatively, we could keep `useBreakpoint` inside utils and call it from `useStyleProps` as you had, but keep the media queries as a singleton that is initialized the first time `useBreakpoint` is called and reused for subsequent calls. We'd need to keep a ref-count of how many listeners there are, and decrement when the component unmounts. When reaching zero we'd destroy the media query.

Either option works for me. Feel free to try them out and see how they work.

Reply:
I like the first approach! Thank you.